### PR TITLE
Remove type casting from dispatch interface

### PR DESCRIPTION
**Pull request by timbussmann**
Removes the need to cast the address type of the `TransportOperation` to a specific implementation to determine whether it's an unicast or a multicast operation.

* IDispatchMessages.Dispatch now takes a `TransportOperations` parameter instead of `IEnumerable<TransportOperation>`.
* `TransportOperations` contains two collections: `UnicastTransportOperations` and `MulticastTransportOperations`.
* `DispatchOptions` has been removed and the properties are directly present on the `Unicast/MulticastTransportOperation`
* for internal usage, you can easily create a `TransportOperations` instance by providing a collection of `TransportOperation` to the ctor. It will split and transform the provided transport operations.

Connects to #3214

### CODE REVIEW

**Comment by timbussmann** [comment]
@danielmarbach @Particular/tf-async-transport does that solve the issues mentioned in #3214?

**Comment by andreasohlund** [comment]
:+1:

**Comment by SeanFeldman** [comment]
Awesome!

File: src/NServiceBus.Core/Transports/UnicastTransportOperation.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+        /// <summary>
+        /// The message to be sent over the transport.
+        /// </summary>
+        public OutgoingMessage Message { get; set; }
+
+        /// <summary>
+        /// Defines the receiver of the message.
+        /// </summary>
+        public UnicastAddressTag AddressTag { get; set; }
```
Comment (johnsimons) [inline src/NServiceBus.Core/Transports/UnicastTransportOperation.cs]: I still fail to see what value `UnicastAddressTag` provides, a consumer will always want access the `Destination` prop, I would just expose a `string Destination` prop instead.

File: src/NServiceBus.Core/Transports/MulticastTransportOperation.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+        /// <summary>
+        /// The message to be sent over the transport.
+        /// </summary>
+        public OutgoingMessage Message { get; }
+
+        /// <summary>
+        /// Defines the receivers of the message.
+        /// </summary>
+        public MulticastAddressTag AddressTag { get; }
```
Comment (johnsimons) [inline src/NServiceBus.Core/Transports/MulticastTransportOperation.cs]: I still fail to see what value `MulticastAddressTag` provides, a consumer will always want access the `MessageType` prop, I would just expose a `Type MessageType` prop instead. Less cluter for the user of the api.

File: src/NServiceBus.Core/Transports/UnicastTransportOperation.cs (hunk 1 above)
Comment (johnsimons) [inline src/NServiceBus.Core/Transports/UnicastTransportOperation.cs]: This has also the benefit of not exposing `UnicastAddressTag` and `MulticastAddressTag` classes as public, the less public api a implementer needs to deal with the best IMO :smile:

Comment (SimonCropp) [inline src/NServiceBus.Core/Transports/UnicastTransportOperation.cs]: Agree with @johnsimons here. While there may be some value in these wrapper classes internally i just see friction forcing them onto users. Considering many of our own team, when seeing them, respond with "huh why, what is the value of this" i think the consumers of our api will have problems groc'ing the concept.

Comment (SeanFeldman) [inline src/NServiceBus.Core/Transports/UnicastTransportOperation.cs]: I see your point @johnsimons. `UnicastAddressTag` has a `Destination` vs `MulticastAddressTag` that exposes `MessageType`. When all operations were in a single collection, it made sense to have the differentiation, but now that we have to distinguished collections perhaps we don't need those tags.

Comment (timbussmann) [inline src/NServiceBus.Core/Transports/UnicastTransportOperation.cs]: @johnsimons I agree with you that `AddressTag` seems to be unnecessary noise here. Going to replace it with a `string` for the unicast and a `Type` for the multicast operation. :+1:

**Comment by timbussmann** [comment]
updated, inlined multicast and unicast address tag properties.

**Comment by danielmarbach** [comment]
Needs a rebase

**Comment by timbussmann** [comment]
@danielmarbach rebased

**Pull request merged**